oxenstored/transaction.ml

```diff
@@ -219,9 +219,76 @@ let rm t perm path =
   add_wop t Xenbus.Xb.Op.Rm path
 
 let ls t perm path =
-  let r = Store.ls t.store perm path in
+  let _node, r = Store.ls t.store perm path in
   set_read_lowpath t path ; r
 
+let ls_partial t perm path directory_cache =
+  let configurable_cache_size = 8 in
+  let check_hashtable_bounds ht =
+    if Hashtbl.length ht >= configurable_cache_size then
+      let key, _min_timestamp =
+        Hashtbl.fold
+          (fun key (_, _, _, timestamp) (min_key, min_timestamp) ->
+            if timestamp < min_timestamp then
+              (Some key, timestamp)
+            else
+              (min_key, min_timestamp)
+          )
+          ht (None, infinity)
+      in
+      Hashtbl.remove ht (Option.get key)
+  in
+
+  (* Return the cached buffer if the node hasn't changed; otherwise
+     create or refresh the cache *)
+  let ( let* ) = Option.bind in
+  let cache_opt = Hashtbl.find_opt directory_cache path in
+  let cache =
+    let* cached_node, gen_count, cache, _timestamp = cache_opt in
+
+    let* cur_node = Store.get_node t.store path in
+    (* Regenerate the cache entry if the underlying node changed *)
+    if cached_node == cur_node then
+      (* TODO check_parents_perms_identical oldroot currentroot path*)
+      Some (cached_node, gen_count, cache)
+    else
+      None
+  in
+  let ret =
+    match cache with
+    | Some (cached_node, gen_count, cache) ->
+        (* Update the timestamp in the cache *)
+        Hashtbl.replace directory_cache path
+          (cached_node, gen_count, cache, Unix.gettimeofday ()) ;
```

Comment:
I'd rather use a monotonic clock here, but it doesn't look like there's one in the standard library :(

Reply:
I don't think we're limited to the standard library anymore now that oxenstored is forked. Maybe I'm missing something?

Reply:
It's good practise to keep the dependency cone as small as it's practical